app/policies/application_policy.rb

```diff
@@ -0,0 +1,95 @@
+class ApplicationPolicy
+  attr_reader :user, :record
+
+  def initialize(user, record)
+    @user = user
+    @record = record
+  end
+
+  @scopes_by_action = {}
+
+  def self.scope(*actions, with:)
+    actions.each do |action|
+      @scopes_by_action ||= {}
+      @scopes_by_action[action] = with
+    end
+  end
+
+  def self.scopes_by_action
+    @scopes_by_action || {}
+  end
+
+  def scope_for(action)
+    scope = if record.is_a?(Class)
+              record
+            else
+              record.class
+            end
+
+    self.class.scopes_by_action[action].new(user, scope).resolve(action)
+  end
+
+  class Scope
+    def self.roles_for_private_scope(roles)
+      @roles ||= []
+      @roles = roles
+    end
+
+    def self.roles
+      @roles || []
+    end
+
+    attr_reader :user, :scope
+
+    def initialize(user, scope)
+      @user = user
+      @scope = scope
+    end
+
+    def resolve(action = nil)
+      case
+      when user.is_admin?
+        scope.all
+      when user.logged_in?
+        user_can_access_scope(
+          private_query(action, user, self.class.roles),
+          public_flag
+        )
+      when public_flag
+        public_scope
+      else
+        scope.none
+      end
+    end
+
+    def public_flag
+      return false unless respond_to?(:public_scope)
+      !public_scope.is_a?(ActiveRecord::NullRelation)
+    end
+
+    def user_can_access_scope(private_query, public_flag)
+      accessible = scope
```

Comment:
does this mutate the `@scope` variable for the instance? I'm wondering about side effects of this method being called after setting up an initial scope.

Reply:
AFAIK all Arel stuff returns new scopes, never mutate?

Reply:
You are correct, there is no mutation happening here. I tested this and the `@scope` instance variable passed in via the `ApplicationPolicy.scopes_by_action` isn't an actual AR scope, it's a class that is used to create the scopes. What confused me was the assignment `accessible = scope` as it's not needed and doesn't relay the intention either. 

Perhaps the `@scope` instance var should really be `@scope_klass` and the it can be used across the `Scope` object to resolve the underlying scopes. E.g. here could be  
```
accessible_scope = scope_klass.where(id: private_query.select(:resource_id))
accessible_scope = accessible_scope.or(public_scope) if public_flag
accessible_scope
```